src/Orleans/EventSourcing/JournaledGrainState.cs

```diff
@@ -28,36 +28,58 @@ MIT License
 
 namespace Orleans.EventSourcing
 {
-    public abstract class JournaledGrainState : GrainState
+    public abstract class JournaledGrainState<TGrainState> : GrainState
+        where TGrainState : JournaledGrainState<TGrainState>
     {
         private List<StateEvent> events = new List<StateEvent>();
-        protected JournaledGrainState(Type type )
-            : base(type.FullName)
+
+        protected JournaledGrainState()
+            : base(typeof(TGrainState).FullName)
         {
         }
 
         public IEnumerable<StateEvent> Events
         {
             get { return events.AsReadOnly(); }
-            set { events = new List<StateEvent>(value); }
         }
 
         public int Version { get; private set; }
 
-        public void AddEvent(StateEvent @event)
+        public void AddEvent<TEvent>(TEvent @event)
+            where TEvent: StateEvent
         {
             events.Add(@event);
-            ApplyEvent(@event);
+
+            DynamicStateTransition(@event);
+            //StaticStateTransition(@event);
+
             Version++;
         }
 
-        public abstract void ApplyEvent(StateEvent @event);
-
         public override void SetAll(IDictionary<string, object> values)
         {
             base.SetAll(values);
-            foreach(StateEvent @event in Events)
-                ApplyEvent(@event);
+
+            dynamic me = this;
+
+            foreach (StateEvent @event in Events)
+                me.Apply(@event);
+        }
+
+        private void DynamicStateTransition<TEvent>(TEvent @event)
+            where TEvent : StateEvent
+        {
+            dynamic me = this;
+            me.Apply(this, @event);
+        }
+
+        private void StaticStateTransition<TEvent>(TEvent @event)
+            where TEvent : StateEvent
+        {
+            var transition = this as IJournaledGrainStateTransition<TGrainState, TEvent>;
+            
```

Comment:
Jason, this is what we called a static approach, which we agreed not to do. The consensus as of now is to go with dynamic approach (see examples for both approaches in this PR and in the tests). Your other comment about potentially fragility of the dynamic approach is a very valid one, which partly bothers us all. What is the concrete suggestion to improve it? How do you do "handler registration" to a (lets not forget **virtual**) grain ?